### PR TITLE
sync on demand: in case of parallel requests pull image just once, cl…

### PR DESCRIPTION
**Pull request by peusebiu**
…oses #344

Signed-off-by: Petu Eusebiu <peusebiu@cisco.com>

**What type of PR is this?**

**Which issue does this PR fix**:

#344 
**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs from IPAMD/CNI showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades. Has updating a running cluster been tested?**:

**Does this change require updates to the CNI daemonset config files to work?**:

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by rchincha** [comment]
Why not solve this with per-image or per-repo channel? Do a lookup, if the image is not found, send the request to the downloader goroutine along with the callback context (http.response, image, ...)
Try not to add too many locks in the code, makes it unnecessarily complex and hard to debug going fwd.

File: pkg/api/routes.go [hunk 1]
```diff
@@ -47,11 +48,12 @@ const (
 )
 
 type RouteHandler struct {
-	c *Controller
+	c              *Controller
+	demandedImages sync.Map
```
Comment (peusebiu) [inline pkg/api/routes.go:52]: Thinking about it, not sure weather sync.Map is overkill. Maybe a normal map with an internal mutex would suffice.

**Comment by peusebiu** [comment]
@rchincha is this better?

basically I have per image channel (map) where I store currently downloading images and their channel, if key is present then wait for sync to finish (for channel to be closed), otherwise sync on demand and remove key after and close the channel.

**Comment by peusebiu** [comment]
closed in f89925fb274172b2664192d4243479c345c5b41b

**Pull request closed without merging**